frmts/grok/jp2dataset.h

```diff
@@ -0,0 +1,149 @@
+/******************************************************************************
+ *
+ * Project:  JPEG2000 driver based on OpenJPEG or Grok library
+ * Purpose:  JPEG2000 driver based on OpenJPEG or Grok library
+ * Author:   Aaron Boxer, <boxerab at protonmail dot com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2010-2014, Even Rouault <even dot rouault at spatialys dot com>
+ * Copyright (c) 2015, European Union (European Environment Agency)
+ * Copyright (c) 2023, Grok Image Compression Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#pragma once
+
+#include "cpl_atomic_ops.h"
+#include "cpl_multiproc.h"
+#include "cpl_string.h"
+#include "cpl_worker_thread_pool.h"
+#include "gdaljp2abstractdataset.h"
+#include "gdaljp2metadata.h"
+
+#include "context.h"
+
+/************************************************************************/
+/* ==================================================================== */
+/*                           JP2Dataset                         */
+/* ==================================================================== */
+/************************************************************************/
+class JP2RasterBand;
+
+class JP2Dataset final : public GDALJP2AbstractDataset, public jp2_context_impl
```

Comment:
I appreciate the effort of trying to factor common codes, but I don't think as done currently it is going to work in practice.
The jp2dataset.h file and jp2dataset.cpp (likely renamed to jp2dataset_impl.hpp) should probably be moved to a frmts/openjpeg_like_common directory, and included by both drivers.  And as the JP2Dataset class is going to be instanciated twice whice with the diffrent context implementation, there should be either some #define trickery to have single name, or using C++ templatating to enure the One Definition Rule isn't violated.

JP2Dataset is quite of a generic name. I'd suggest JP2OJPLikeDataset to reflect that it applies to OpenJPEG like APIs.

Reply:
Thanks for the review. I've reworked things using templates and renamed to JP2OPJLikeDataset, so it can now be easily reused by both drivers. For this PR,  I would prefer to not change JP2OpenJPEG and focus on getting everything working for the new driver. We can leave that refactor for another PR.